### ISSUE TITLE
With CD audio, intro doesn't play, and E1M1 plays instead of intermission

### ISSUE THREAD

**Issue by rfmerrill** [viciious/d32xr#84]
This is with 3.1

I have the intermission as track 28 and the intro as track 29, per the guide in the wiki. As the title says, I get a silent intro, and when I complete a level, e1m1 plays instead of the intermission music.

I've also verified I can play both of these tracks through the MegaCD's CD player menu, so it's able to read and play them.

**Comment by rfmerrill** [comment]
I just booted it up again and e1m1 played for the intro as well

**Comment by rfmerrill** [comment]
Ah. Is there an up to date tracklist to follow somewhere?

**Comment by viciious** [comment]
Each map should have its own dedicated audio track, followed by:
- the end track
- victory track
- intermission screen track
- the title screen/introduction track

**Comment by rfmerrill** [comment]
That sounds like what I have. How many tracks total should that be? I have 29

**Comment by viciious** [comment]
v3.1 lists 27 maps in its "New game" setup menu screen

**Comment by rfmerrill** [comment]
Gotcha. So I should do e1m1-e1m9, e2m1-e2m9, e3m1-e3m9 and then bunny, victory, inter, intro?

**Comment by rfmerrill** [comment]
Ah ok, I think it's in the order they're listed in the menu. So 
- 1-8 for each episode
- then e1m9, e2m9 
- hell gate (this is a console-only level?)
- Then the last four

**Comment by viciious** [comment]
Audio tracks sequencing should follow the 'mapNumber' field ordering here: https://github.com/viciious/d32xr/blob/master/content/files/dmapinfo

**Comment by rfmerrill** [comment]
Thank you very much for your helpful responses, that helps a lot!

Unfortunately I burned a new disc with 31 tracks (27 + 4), and while every map I start on has the music that I expect it to, I still get e1m1 for both the title screen and intermission :(

I should get a build environment going so I can see if I can't debug this myself.

**Comment by rfmerrill** [comment]
Since I don't have much else to go on I decided to try other hardware to make sure it wasn't revision or unit dependent.

Original setup: JP 32X, JP Model 1 VA6 mega drive, JP Model 1 Mega CD
Second setup: JP 32X, US Model 2 VA1 mega drive, US Model 2 Sega CD (Funai)

Same behavior in both cases.

**Comment by rfmerrill** [comment]
@ChillyWillyGuru It's a rare edge case, but what happens if a CD actually *does* have >=64 tracks? I know of at least one commercially pressed CD that has 99 (Fat Wreck Chords' "Short Music For Short People").

Of course there's not necessarily a reason for D32XR to actually support that, but the error handling should ideally be graceful :)

**Comment by rfmerrill** [comment]
Earlier I tested this patch and confirmed it fixed the bug:

```
diff --git a/marshw.c b/marshw.c
index 6f6de23..9048ebe 100644
--- a/marshw.c
+++ b/marshw.c
@@ -305,7 +305,9 @@ void Mars_UpdateCD(void)
        MARS_SYS_COMM0 = 0x0600;
        while (MARS_SYS_COMM0);
        mars_cd_ok = MARS_SYS_COMM2;
-       mars_num_cd_tracks = mars_cd_ok >> 2;
+       // XXX FIXME: unsure how this ends up with bit 6
+       //            erroneously set somehow.
+       mars_num_cd_tracks = (mars_cd_ok >> 2) & 0x3F;
        mars_cd_ok = mars_cd_ok & 0x3;
 }
 ```

**Comment by rfmerrill** [comment]
@ChillyWillyGuru OK I think I see the issue. For my own education, can you tell me if I'm right?

- `InitCD` returns 1 if it succeds, which sets the `cd_ok` to `0x0001`
- If `read_cdstate` sees a disk present, it does `move.b #1, cd_ok` which (since we're on a big-endian CPU) sets `cd_ok` to `0x0101`
- The result delivered to the SH-1 is `(num_tracks << 2) | cd_ok` and the SH-1 expects to see cd_ok in bits 0 and 1, not 0 and 5. This is the part where there's a mixup.

**Comment by rfmerrill** [comment]
Thanks for the clarification!

So if I'm reading this right, the 1 in the upper byte tells us if a disc is physically present and ready in the CD player. This might be redundant information to the SH-1, since it should be able to tell that state from the number of tracks (which will be zero if there's no disc). Did I get that?

**Comment by viciious** [comment]
Fixed in 84f12290f43c4aa6dc4794c8c1c7fbd788747afd
